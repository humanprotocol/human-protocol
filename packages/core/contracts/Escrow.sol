// SPDX-License-Identifier: MIT

pragma solidity >=0.6.2;

import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
import '@openzeppelin/contracts/security/ReentrancyGuard.sol';

import './interfaces/IRewardPool.sol';
import './interfaces/IEscrow.sol';
import './utils/SafeMath.sol';

contract Escrow is IEscrow, ReentrancyGuard {
    using SafeMath for uint256;

    bytes4 private constant FUNC_SELECTOR_BALANCE_OF =
        bytes4(keccak256('balanceOf(address)'));
    bytes4 private constant FUNC_SELECTOR_TRANSFER =
        bytes4(keccak256('transfer(address,uint256)'));

    string constant ERROR_ZERO_ADDRESS = 'Escrow: zero address';

    event TrustedHandlerAdded(address _handler);
    event IntermediateStorage(string _url, string _hash);
    event Pending(string manifest, string hash);
    event BulkTransfer(uint256 indexed _txId, uint256 _bulkCount);
    event Cancelled();
    event Completed();

    EscrowStatuses public override status;

    address public reputationOracle;
    address public recordingOracle;
    address public launcher;
    address payable public canceler;

    uint256 public reputationOracleStake;
    uint256 public recordingOracleStake;
    uint256 private constant BULK_MAX_VALUE = 1e9 * (10 ** 18);
    uint32 private constant BULK_MAX_COUNT = 100;

    address public token;

    string public manifestUrl;
    string public manifestHash;
    uint256 public remainingSolutions;

    string public finalResultsUrl;
    string public finalResultsHash;

    uint256 public duration;

    uint256[] public finalAmounts;
    bool public bulkPaid;

    mapping(address => bool) public areTrustedHandlers;

    constructor(
        address _token,
        address payable _canceler,
        uint256 _duration,
        address[] memory _handlers
    ) {
        require(_token != address(0), ERROR_ZERO_ADDRESS);
        require(_canceler != address(0), ERROR_ZERO_ADDRESS);

        token = _token;
        status = EscrowStatuses.Launched;
        duration = _duration.add(block.timestamp); // solhint-disable-line not-rely-on-time
        launcher = msg.sender;
        canceler = _canceler;
        areTrustedHandlers[_canceler] = true;
        areTrustedHandlers[msg.sender] = true;
        addTrustedHandlers(_handlers);
    }

    function getBalance() public view returns (uint256) {
        (bool success, bytes memory returnData) = token.staticcall(
            abi.encodeWithSelector(FUNC_SELECTOR_BALANCE_OF, address(this))
        );
        if (success) {
            return abi.decode(returnData, (uint256));
        }
        return 0;
    }

    function addTrustedHandlers(address[] memory _handlers) public {
        require(
            areTrustedHandlers[msg.sender],
            'Address calling cannot add trusted handlers'
        );
        for (uint256 i = 0; i < _handlers.length; i++) {
            require(_handlers[i] != address(0), ERROR_ZERO_ADDRESS);
            areTrustedHandlers[_handlers[i]] = true;
            emit TrustedHandlerAdded(_handlers[i]);
        }
    }

    // The escrower puts the Token in the contract without an agentless
    // and assigsn a reputation oracle to payout the bounty of size of the
    // amount specified
    function setup(
        address _reputationOracle,
        address _recordingOracle,
        uint256 _reputationOracleStake,
        uint256 _recordingOracleStake,
        string memory _url,
        string memory _hash,
        uint256 _solutionsRequested
    ) external trusted notExpired {
        require(
            _reputationOracle != address(0),
            'Invalid or missing token spender'
        );
        require(
            _recordingOracle != address(0),
            'Invalid or missing token spender'
        );
        require(_solutionsRequested > 0, 'Invalid or missing solutions');
        uint256 totalStake = _reputationOracleStake.add(_recordingOracleStake);
        require(totalStake <= 100, 'Stake out of bounds');
        require(
            status == EscrowStatuses.Launched,
            'Escrow not in Launched status state'
        );

        reputationOracle = _reputationOracle;
        recordingOracle = _recordingOracle;
        areTrustedHandlers[reputationOracle] = true;
        areTrustedHandlers[recordingOracle] = true;

        reputationOracleStake = _reputationOracleStake;
        recordingOracleStake = _recordingOracleStake;

        manifestUrl = _url;
        manifestHash = _hash;
        remainingSolutions = _solutionsRequested;
        status = EscrowStatuses.Pending;
        emit Pending(manifestUrl, manifestHash);
    }

    function abort() external trusted notComplete notPaid {
        if (getBalance() != 0) {
            cancel();
        }
        selfdestruct(canceler);
    }

    function cancel()
        public
        trusted
        notBroke
        notComplete
        notPaid
        nonReentrant
        returns (bool)
    {
        _safeTransfer(canceler, getBalance());
        status = EscrowStatuses.Cancelled;
        emit Cancelled();
        return true;
    }

    function complete() external notExpired {
        require(
            areTrustedHandlers[msg.sender],
            'Address calling is not trusted'
        );
        require(status == EscrowStatuses.Paid, 'Escrow not in Paid state');
        status = EscrowStatuses.Complete;
        emit Completed();
    }

    function storeResults(
        string memory _url,
        string memory _hash
    ) external trusted notExpired {
        require(
            status == EscrowStatuses.Pending ||
                status == EscrowStatuses.Partial,
            'Escrow not in Pending or Partial status state'
        );
        _storeResult(_url, _hash);
        emit IntermediateStorage(_url, _hash);
    }

    function bulkPayOut(
        address[] memory _recipients,
        uint256[] memory _amounts,
        string memory _url,
        string memory _hash,
        uint256 _txId
    )
        external
        trusted
        notBroke
        notLaunched
        notPaid
        notExpired
        nonReentrant
        returns (bool)
    {
        require(
            _recipients.length == _amounts.length,
            "Amount of recipients and values don't match"
        );
        require(_recipients.length < BULK_MAX_COUNT, 'Too many recipients');
        require(
            status != EscrowStatuses.Complete &&
                status != EscrowStatuses.Cancelled,
            'Invalid status'
        );

        uint256 balance = getBalance();
        bulkPaid = false;
        uint256 aggregatedBulkAmount = 0;
        for (uint256 i; i < _amounts.length; i++) {
            aggregatedBulkAmount = aggregatedBulkAmount.add(_amounts[i]);
        }
        require(aggregatedBulkAmount < BULK_MAX_VALUE, 'Bulk value too high');

        if (balance < aggregatedBulkAmount) {
            return bulkPaid;
        }

        _storeResult(_url, _hash);

        (
            uint256 reputationOracleFee,
            uint256 recordingOracleFee
        ) = finalizePayouts(_amounts);

        for (uint256 i = 0; i < _recipients.length; ++i) {
            uint256 amount = finalAmounts[i];
            if (amount == 0) {
                continue;
            }
            finalAmounts[i] = 0;
            _safeTransfer(_recipients[i], amount);
        }

        delete finalAmounts;

        _safeTransfer(reputationOracle, reputationOracleFee);
        _safeTransfer(recordingOracle, recordingOracleFee);

        bulkPaid = true;
        balance = getBalance();
        if (bulkPaid) {
            if (status == EscrowStatuses.Pending) {
                status = EscrowStatuses.Partial;
                remainingSolutions = remainingSolutions.sub(_recipients.length);
            }
            if (
                balance > 0 &&
                status == EscrowStatuses.Partial &&
                remainingSolutions == 0
            ) {
                _safeTransfer(canceler, balance);
                status = EscrowStatuses.Paid;
            }
            if (balance == 0 && status == EscrowStatuses.Partial) {
                status = EscrowStatuses.Paid;
            }
        }
        emit BulkTransfer(_txId, _recipients.length);
        return true;
    }

    function finalizePayouts(
        uint256[] memory _amounts
    ) internal returns (uint256, uint256) {
        uint256 reputationOracleFee = 0;
        uint256 recordingOracleFee = 0;
        for (uint256 j; j < _amounts.length; j++) {
            uint256 singleReputationOracleFee = reputationOracleStake
                .mul(_amounts[j])
                .div(100);
            uint256 singleRecordingOracleFee = recordingOracleStake
                .mul(_amounts[j])
                .div(100);
            uint256 amount = _amounts[j].sub(singleReputationOracleFee).sub(
                singleRecordingOracleFee
            );
            reputationOracleFee = reputationOracleFee.add(
                singleReputationOracleFee
            );
            recordingOracleFee = recordingOracleFee.add(
                singleRecordingOracleFee
            );
            finalAmounts.push(amount);
        }
        return (reputationOracleFee, recordingOracleFee);
    }

    function _safeTransfer(address to, uint256 value) internal {
        SafeERC20.safeTransfer(IERC20(token), to, value);
    }

    function _storeResult(string memory _url, string memory _hash) internal {
        bool writeOnchain = bytes(_hash).length != 0 || bytes(_url).length != 0;
        if (writeOnchain) {
            // Be sure both of them are not zero
            finalResultsUrl = _url;
            finalResultsHash = _hash;
        }
    }

    modifier trusted() {
        require(areTrustedHandlers[msg.sender], 'Address calling not trusted');
        _;
    }

    modifier notBroke() {
        require(getBalance() != 0, 'Token contract out of funds');
        _;
    }

    modifier notComplete() {
        require(
            status != EscrowStatuses.Complete,
            'Escrow in Complete status state'
        );
        _;
    }

    modifier notPaid() {
        require(status != EscrowStatuses.Paid, 'Escrow in Paid status state');
        _;
    }

    modifier notLaunched() {
        require(
            status != EscrowStatuses.Launched,
            'Escrow in Launched status state'
        );
        _;
    }

    modifier notExpired() {
        require(duration > block.timestamp, 'Contract expired'); // solhint-disable-line not-rely-on-time
        _;
    }
}
