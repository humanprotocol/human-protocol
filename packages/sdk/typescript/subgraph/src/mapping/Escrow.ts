import {
  BulkTransfer,
  Cancelled,
  Completed,
  Escrow as EscrowContract,
  IntermediateStorage,
  Pending,
} from '../../generated/templates/Escrow/Escrow';
import { LegacyEscrow as LegacyEscrowContract } from '../../generated/templates/Escrow/LegacyEscrow';
import {
  BulkPayoutEvent,
  CancelledStatusEvent,
  CompletedStatusEvent,
  Escrow,
  EscrowStatistics,
  PaidStatusEvent,
  PartialStatusEvent,
  PendingStatusEvent,
  SetupEvent,
  StoreResultsEvent,
  Worker,
} from '../../generated/schema';
import { Address, BigInt, dataSource } from '@graphprotocol/graph-ts';
import { ZERO_BI, ONE_BI } from './utils/number';
import { toEventId } from './utils/event';
import { getEventDayData } from './utils/dayUpdates';

export const STATISTICS_ENTITY_ID = 'escrow-statistics-id';

function constructStatsEntity(): EscrowStatistics {
  const entity = new EscrowStatistics(STATISTICS_ENTITY_ID);

  entity.fundEventCount = ZERO_BI;
  entity.setupEventCount = ZERO_BI;
  entity.storeResultsEventCount = ZERO_BI;
  entity.bulkPayoutEventCount = ZERO_BI;
  entity.pendingStatusEventCount = ZERO_BI;
  entity.cancelledStatusEventCount = ZERO_BI;
  entity.partialStatusEventCount = ZERO_BI;
  entity.paidStatusEventCount = ZERO_BI;
  entity.completedStatusEventCount = ZERO_BI;
  entity.totalEventCount = ZERO_BI;
  entity.totalEscrowCount = ZERO_BI;

  return entity;
}

export function createOrLoadEscrowStatistics(): EscrowStatistics {
  let statsEntity = EscrowStatistics.load(STATISTICS_ENTITY_ID);

  if (!statsEntity) {
    statsEntity = constructStatsEntity();
  }

  return statsEntity;
}

export function createOrLoadWorker(address: Address): Worker {
  let worker = Worker.load(address.toHex());

  if (!worker) {
    worker = new Worker(address.toHex());
    worker.address = address;
    worker.totalAmountReceived = ZERO_BI;
    worker.payoutCount = ZERO_BI;
  }

  return worker;
}

export function handlePending(event: Pending): void {
  // Create SetupEvent entity
  const setupEventEntity = new SetupEvent(toEventId(event));
  setupEventEntity.block = event.block.number;
  setupEventEntity.timestamp = event.block.timestamp;
  setupEventEntity.txHash = event.transaction.hash;
  setupEventEntity.escrowAddress = event.address;
  setupEventEntity.sender = event.transaction.from;
  setupEventEntity.save();

  // Create PendingStatusEvent entity
  const statusEventEntity = new PendingStatusEvent(toEventId(event));
  statusEventEntity.block = event.block.number;
  statusEventEntity.timestamp = event.block.timestamp;
  statusEventEntity.txHash = event.transaction.hash;
  statusEventEntity.escrowAddress = event.address;
  statusEventEntity.sender = event.transaction.from;
  statusEventEntity.save();

  // Updates escrow statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.setupEventCount = statsEntity.setupEventCount.plus(ONE_BI);
  statsEntity.pendingStatusEventCount =
    statsEntity.pendingStatusEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(
    ONE_BI.plus(ONE_BI)
  );
  statsEntity.save();

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailySetupEventCount =
    eventDayData.dailySetupEventCount.plus(ONE_BI);
  eventDayData.dailyPendingStatusEventCount =
    eventDayData.dailyPendingStatusEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount = eventDayData.dailyTotalEventCount.plus(
    ONE_BI.plus(ONE_BI)
  );

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    escrowEntity.manifestUrl = event.params.manifest;
    escrowEntity.manifestHash = event.params.hash;
    escrowEntity.status = 'Pending';

    // Read data on-chain
    const escrowContract = EscrowContract.bind(event.address);
    const legacyEscrowContract = LegacyEscrowContract.bind(event.address);

    // Reputation & Recording Oracle Fee Variable is changed over time
    // For old one, it was oracleStake, for new one it is oracleFeePercentage

    const reputationOracle = escrowContract.try_reputationOracle();
    if (!reputationOracle.reverted) {
      escrowEntity.reputationOracle = reputationOracle.value;
    }
    const reputationOracleFeePercentage =
      escrowContract.try_reputationOracleFeePercentage();
    if (!reputationOracleFeePercentage.reverted) {
      escrowEntity.reputationOracleFee = BigInt.fromI32(
        reputationOracleFeePercentage.value
      );
    }
    const reputationOracleStake =
      legacyEscrowContract.try_reputationOracleStake();
    if (!reputationOracleStake.reverted) {
      escrowEntity.reputationOracleFee = reputationOracleStake.value;
    }

    const recordingOracle = escrowContract.try_recordingOracle();
    if (!recordingOracle.reverted) {
      escrowEntity.recordingOracle = recordingOracle.value;
    }
    const recordingOracleFeePercentage =
      escrowContract.try_recordingOracleFeePercentage();
    if (!recordingOracleFeePercentage.reverted) {
      escrowEntity.recordingOracleFee = BigInt.fromI32(
        recordingOracleFeePercentage.value
      );
    }
    const recordingOracleStake =
      legacyEscrowContract.try_recordingOracleStake();
    if (!recordingOracleStake.reverted) {
      escrowEntity.recordingOracleFee = recordingOracleStake.value;
    }

    const exchangeOracle = escrowContract.try_exchangeOracle();
    if (!exchangeOracle.reverted) {
      escrowEntity.exchangeOracle = exchangeOracle.value;
    }
    const exchangeOracleFeePercentage =
      escrowContract.try_exchangeOracleFeePercentage();
    if (!exchangeOracleFeePercentage.reverted) {
      escrowEntity.exchangeOracleFee = BigInt.fromI32(
        exchangeOracleFeePercentage.value
      );
    }

    escrowEntity.save();
  }
}

export function handleIntermediateStorage(event: IntermediateStorage): void {
  // Create StoreResultsEvent entity
  const eventEntity = new StoreResultsEvent(toEventId(event));
  eventEntity.block = event.block.number;
  eventEntity.timestamp = event.block.timestamp;
  eventEntity.txHash = event.transaction.hash;
  eventEntity.escrowAddress = event.address;
  eventEntity.sender = event.transaction.from;
  eventEntity.intermediateResultsUrl = event.params._url;
  eventEntity.save();

  // Updates escrow statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.storeResultsEventCount =
    statsEntity.storeResultsEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);
  statsEntity.save();

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailyStoreResultsEventCount =
    eventDayData.dailyStoreResultsEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount =
    eventDayData.dailyTotalEventCount.plus(ONE_BI);
  eventDayData.save();

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    escrowEntity.intermediateResultsUrl = event.params._url;
    escrowEntity.save();
  }
}

export function handleBulkTransfer(event: BulkTransfer): void {
  // Create BulkPayoutEvent entity
  const eventEntity = new BulkPayoutEvent(toEventId(event));
  eventEntity.block = event.block.number;
  eventEntity.timestamp = event.block.timestamp;
  eventEntity.txHash = event.transaction.hash;
  eventEntity.escrowAddress = event.address;
  eventEntity.sender = event.transaction.from;
  eventEntity.bulkPayoutTxId = event.params._txId;
  eventEntity.bulkCount = BigInt.fromI32(event.params._recipients.length);
  eventEntity.save();

  // Update escrow statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.bulkPayoutEventCount =
    statsEntity.bulkPayoutEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailyBulkPayoutEventCount =
    eventDayData.dailyBulkPayoutEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount =
    eventDayData.dailyTotalEventCount.plus(ONE_BI);

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    escrowEntity.status = event.params._isPartial ? 'Partial' : 'Paid';

    // Read data on-chain
    const escrowContract = EscrowContract.bind(event.address);
    const finalResultsUrl = escrowContract.try_finalResultsUrl();
    if (!finalResultsUrl.reverted) {
      escrowEntity.finalResultsUrl = finalResultsUrl.value;
    }

    escrowEntity.save();
  }

  if (event.params._isPartial) {
    const statusEventEntity = new PartialStatusEvent(toEventId(event));
    statusEventEntity.block = event.block.number;
    statusEventEntity.timestamp = event.block.timestamp;
    statusEventEntity.txHash = event.transaction.hash;
    statusEventEntity.escrowAddress = event.address;
    statusEventEntity.sender = event.transaction.from;
    statusEventEntity.save();

    statsEntity.partialStatusEventCount =
      statsEntity.partialStatusEventCount.plus(ONE_BI);
    statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);

    eventDayData.dailyPartialStatusEventCount =
      eventDayData.dailyPartialStatusEventCount.plus(ONE_BI);
    eventDayData.dailyTotalEventCount =
      eventDayData.dailyTotalEventCount.plus(ONE_BI);
  } else {
    const statusEventEntity = new PaidStatusEvent(toEventId(event));
    statusEventEntity.block = event.block.number;
    statusEventEntity.timestamp = event.block.timestamp;
    statusEventEntity.txHash = event.transaction.hash;
    statusEventEntity.escrowAddress = event.address;
    statusEventEntity.sender = event.transaction.from;
    statusEventEntity.save();

    statsEntity.paidStatusEventCount =
      statsEntity.paidStatusEventCount.plus(ONE_BI);
    statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);

    eventDayData.dailyPaidStatusEventCount =
      eventDayData.dailyPaidStatusEventCount.plus(ONE_BI);
    eventDayData.dailyTotalEventCount =
      eventDayData.dailyTotalEventCount.plus(ONE_BI);
  }

  // Save statistics, and event day data
  statsEntity.save();
  eventDayData.save();
}

export function handleCancelled(event: Cancelled): void {
  // Create CancelledStatusEvent entity
  const eventEntity = new CancelledStatusEvent(toEventId(event));
  eventEntity.block = event.block.number;
  eventEntity.timestamp = event.block.timestamp;
  eventEntity.txHash = event.transaction.hash;
  eventEntity.escrowAddress = event.address;
  eventEntity.sender = event.transaction.from;
  eventEntity.save();

  // Update statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.cancelledStatusEventCount =
    statsEntity.cancelledStatusEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);
  statsEntity.save();

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailyCancelledStatusEventCount =
    eventDayData.dailyCancelledStatusEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount =
    eventDayData.dailyTotalEventCount.plus(ONE_BI);
  eventDayData.save();

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    escrowEntity.status = 'Cancelled';
    escrowEntity.save();
  }
}

export function handleCompleted(event: Completed): void {
  // Create CompletedStatusEvent entity
  const eventEntity = new CompletedStatusEvent(toEventId(event));
  eventEntity.block = event.block.number;
  eventEntity.timestamp = event.block.timestamp;
  eventEntity.txHash = event.transaction.hash;
  eventEntity.escrowAddress = event.address;
  eventEntity.sender = event.transaction.from;
  eventEntity.save();

  // Update statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.completedStatusEventCount =
    statsEntity.completedStatusEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);
  statsEntity.save();

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailyCompletedStatusEventCount =
    eventDayData.dailyCompletedStatusEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount =
    eventDayData.dailyTotalEventCount.plus(ONE_BI);
  eventDayData.save();

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    escrowEntity.status = 'Complete';
    escrowEntity.save();
  }
}
