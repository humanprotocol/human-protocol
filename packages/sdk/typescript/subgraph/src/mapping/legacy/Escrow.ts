import { dataSource } from '@graphprotocol/graph-ts';

import {
  BulkTransfer,
  Escrow as EscrowContract,
  IntermediateStorage,
  Pending,
} from '../../../generated/templates/LegacyEscrow/Escrow';
import {
  BulkPayoutEvent,
  Escrow,
  EscrowStatusEvent,
  SetupEvent,
  StoreResultsEvent,
} from '../../../generated/schema';
import { createOrLoadEscrowStatistics } from '../Escrow';
import { ONE_BI } from '../utils/number';
import { toEventId } from '../utils/event';
import { getEventDayData } from '../utils/dayUpdates';
import { createTransaction } from '../utils/transaction';

enum EscrowStatuses {
  Launched,
  Pending,
  Partial,
  Paid,
  Complete,
  Cancelled,
}

export function handlePending(event: Pending): void {
  createTransaction(event, 'setup');
  // Create SetupEvent entity
  const setupEventEntity = new SetupEvent(toEventId(event));
  setupEventEntity.block = event.block.number;
  setupEventEntity.timestamp = event.block.timestamp;
  setupEventEntity.txHash = event.transaction.hash;
  setupEventEntity.escrowAddress = event.address;
  setupEventEntity.sender = event.transaction.from;
  setupEventEntity.save();

  // Create EscrowStatusEvent entity
  const statusEventEntity = new EscrowStatusEvent(toEventId(event));
  statusEventEntity.block = event.block.number;
  statusEventEntity.timestamp = event.block.timestamp;
  statusEventEntity.txHash = event.transaction.hash;
  statusEventEntity.escrowAddress = event.address;
  statusEventEntity.sender = event.transaction.from;
  statusEventEntity.status = 'Pending';

  // Updates escrow statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.setupEventCount = statsEntity.setupEventCount.plus(ONE_BI);
  statsEntity.pendingStatusEventCount =
    statsEntity.pendingStatusEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(
    ONE_BI.plus(ONE_BI)
  );
  statsEntity.save();

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailySetupEventCount =
    eventDayData.dailySetupEventCount.plus(ONE_BI);
  eventDayData.dailyPendingStatusEventCount =
    eventDayData.dailyPendingStatusEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount = eventDayData.dailyTotalEventCount.plus(
    ONE_BI.plus(ONE_BI)
  );

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    escrowEntity.manifestUrl = event.params.manifest;
    escrowEntity.manifestHash = event.params.hash;
    escrowEntity.status = 'Pending';

    // Read data on-chain
    const escrowContract = EscrowContract.bind(event.address);

    const reputationOracle = escrowContract.try_reputationOracle();
    if (!reputationOracle.reverted) {
      escrowEntity.reputationOracle = reputationOracle.value;
    }
    const reputationOracleStake = escrowContract.try_reputationOracleStake();
    if (!reputationOracleStake.reverted) {
      escrowEntity.reputationOracleFee = reputationOracleStake.value;
    }

    const recordingOracle = escrowContract.try_recordingOracle();
    if (!recordingOracle.reverted) {
      escrowEntity.recordingOracle = recordingOracle.value;
    }
    const recordingOracleStake = escrowContract.try_recordingOracleStake();
    if (!recordingOracleStake.reverted) {
      escrowEntity.recordingOracleFee = recordingOracleStake.value;
    }

    escrowEntity.save();
    statusEventEntity.launcher = escrowEntity.launcher;
  }
  statusEventEntity.save();
}

export function handleIntermediateStorage(event: IntermediateStorage): void {
  createTransaction(event, 'storeResults');
  // Create StoreResultsEvent entity
  const eventEntity = new StoreResultsEvent(toEventId(event));
  eventEntity.block = event.block.number;
  eventEntity.timestamp = event.block.timestamp;
  eventEntity.txHash = event.transaction.hash;
  eventEntity.escrowAddress = event.address;
  eventEntity.sender = event.transaction.from;
  eventEntity.intermediateResultsUrl = event.params._url;
  eventEntity.save();

  // Updates escrow statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.storeResultsEventCount =
    statsEntity.storeResultsEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);
  statsEntity.save();

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailyStoreResultsEventCount =
    eventDayData.dailyStoreResultsEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount =
    eventDayData.dailyTotalEventCount.plus(ONE_BI);
  eventDayData.save();

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    escrowEntity.intermediateResultsUrl = event.params._url;
    escrowEntity.save();
  }
}

export function handleBulkTransfer(event: BulkTransfer): void {
  createTransaction(event, 'bulkTransfer');
  // Create BulkPayoutEvent entity
  const eventEntity = new BulkPayoutEvent(toEventId(event));
  eventEntity.block = event.block.number;
  eventEntity.timestamp = event.block.timestamp;
  eventEntity.txHash = event.transaction.hash;
  eventEntity.escrowAddress = event.address;
  eventEntity.sender = event.transaction.from;
  eventEntity.bulkPayoutTxId = event.params._txId;
  eventEntity.bulkCount = event.params._bulkCount;
  eventEntity.save();

  // Update escrow statistics
  const statsEntity = createOrLoadEscrowStatistics();
  statsEntity.bulkPayoutEventCount =
    statsEntity.bulkPayoutEventCount.plus(ONE_BI);
  statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);

  // Update event day data
  const eventDayData = getEventDayData(event);
  eventDayData.dailyBulkPayoutEventCount =
    eventDayData.dailyBulkPayoutEventCount.plus(ONE_BI);
  eventDayData.dailyTotalEventCount =
    eventDayData.dailyTotalEventCount.plus(ONE_BI);

  // Update escrow entity
  const escrowEntity = Escrow.load(dataSource.address().toHex());
  if (escrowEntity) {
    // Read data on-chain
    const escrowContract = EscrowContract.bind(event.address);
    const escrowStatus = escrowContract.try_status();

    if (!escrowStatus.reverted) {
      // Create EscrowStatusEvent entity
      const statusEventEntity = new EscrowStatusEvent(toEventId(event));
      statusEventEntity.block = event.block.number;
      statusEventEntity.timestamp = event.block.timestamp;
      statusEventEntity.txHash = event.transaction.hash;
      statusEventEntity.escrowAddress = event.address;
      statusEventEntity.sender = event.transaction.from;
      statusEventEntity.launcher = escrowEntity.launcher;
      if (escrowStatus.value == EscrowStatuses.Partial) {
        // Partially Paid Status
        escrowEntity.status = 'Partially Paid';
        statusEventEntity.status = 'Partial';

        statsEntity.partialStatusEventCount =
          statsEntity.partialStatusEventCount.plus(ONE_BI);
        statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);

        eventDayData.dailyPartialStatusEventCount =
          eventDayData.dailyPartialStatusEventCount.plus(ONE_BI);
        eventDayData.dailyTotalEventCount =
          eventDayData.dailyTotalEventCount.plus(ONE_BI);
      } else if (escrowStatus.value == EscrowStatuses.Paid) {
        // Paid Status
        escrowEntity.status = 'Paid';
        statusEventEntity.status = 'Paid';

        statsEntity.paidStatusEventCount =
          statsEntity.paidStatusEventCount.plus(ONE_BI);
        statsEntity.totalEventCount = statsEntity.totalEventCount.plus(ONE_BI);

        eventDayData.dailyPaidStatusEventCount =
          eventDayData.dailyPaidStatusEventCount.plus(ONE_BI);
        eventDayData.dailyTotalEventCount =
          eventDayData.dailyTotalEventCount.plus(ONE_BI);
      }
      statusEventEntity.save();
    }

    const finalResultsUrl = escrowContract.try_finalResultsUrl();
    if (!finalResultsUrl.reverted) {
      escrowEntity.finalResultsUrl = finalResultsUrl.value;
    }
    escrowEntity.save();
  }

  // Save statistics, and event day data
  statsEntity.save();
  eventDayData.save();
}
