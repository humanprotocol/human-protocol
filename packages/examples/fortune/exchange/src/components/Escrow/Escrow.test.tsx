import renderer from 'react-test-renderer';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Escrow } from './Escrow';
import sendFortune from '../../services/RecordingOracle/RecordingClient'

describe('When rendered Escrow component', () => {

    it('renders landing page', async () => {
        render(<Escrow />);
        expect(screen.getByTestId("escrowAddress")).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Confirm" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Send Fortune" })).toBeInTheDocument();
    });


    it('should match snapshot', () => {
        const component = renderer.create(
            <Escrow />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        return;
    });
});