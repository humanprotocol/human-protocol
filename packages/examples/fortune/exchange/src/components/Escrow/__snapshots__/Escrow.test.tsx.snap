// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When rendered Escrow component should match snapshot 1`] = `
<div
  className="escrow-container"
>
  <div
    className="escrow-view"
  >
    <div>
      <input
        data-testid="escrowAddress"
        onChange={[Function]}
        value=""
      />
      <button
        onClick={[Function]}
        type="button"
      >
         Confirm 
      </button>
    </div>
    <span>
       Fill the exchange address to pass the fortune to the recording oracle
    </span>
    <span>
       
      <b>
        Address: 
      </b>
       
       
    </span>
    <span>
       
      <b>
        Status: 
      </b>
       
    </span>
    <span>
       
      <b>
        Balance: 
      </b>
       
    </span>
    <div>
      <input
        onChange={[Function]}
      />
      <button
        onClick={[Function]}
        type="button"
      >
         Send Fortune 
      </button>
    </div>
  </div>
</div>
`;
